s4e-backend/src/main/java/pl/cyfronet/s4e/service/SceneService.java

```diff
@@ -90,11 +111,21 @@ public void deleteProductScenes(Long productId) throws NotFoundException {
         return dates;
     }
 
-    public <T> Optional<T> getMostRecentScene(Long productId, Class<T> projection) throws NotFoundException {
-        if (!productRepository.existsById(productId)) {
-            throw constructNFE("Product", productId);
-        }
+    @Transactional(readOnly = true)
+    public <T extends ProjectionWithId> Optional<T> getMostRecentScene(Long productId, AppUserDetails userDetails, Class<T> projection) throws NotFoundException {
+        ProductProjection product = productRepository.findById(productId, ProductProjection.class)
+                .orElseThrow(() -> constructNFE("Product", productId));
+
         Sort sort = Sort.by(Sort.Order.desc("timestamp"), Sort.Order.asc("id"));
+        if (Product.AccessType.EUMETSAT.equals(product.getAccessType())) {
+            // I stream all the Scenes in chunks of 4 (see query hint in repository) and stop when first matching
+            // is encountered. With 15 minute temporal resolution rarely more than one DB query will have to be
+            // executed as there will be 4 scenes per hour and one of them will be on the hour.
```

Comment:
Okay, this might be a little convoluted, I'll rephrase it like this:
```
Most EUMETSAT products have 15 min temporal resoultion, this means 4 scenes per hour: on 0, 15, 30 and 45 min.
Then, fetching 4 scenes will always yield at least one on the hour scene (open to all) and only a single chunk will be requested.
```

Reply:
Thanks. It's a lot more clear in such a way.